### ISSUE TITLE
Javascript array template not fully applied

### ISSUE THREAD

**Issue by nshmyrev** [swig/swig#865]
The declaration part of the fragment of arrays_javascript.i is not fully applied:

```
%typemap(in, fragment="SWIG_JSCGetIntProperty") int[], int[ANY]
    (int length = 0, v8::Local<v8::Array> array, v8::Local<v8::Value> jsvalue, int i = 0, int res = 0, $*1_ltype temp) {
  if ($input->IsArray())
```

This is a code it generates:

```
static SwigV8ReturnValue _wrap_AudioIndex_lookup(const SwigV8Arguments &args) {
  SWIGV8_HANDLESCOPE();

  v8::Handle<v8::Value> jsresult;
  AudioIndex *arg1 = (AudioIndex *) 0 ;
  unsigned int *arg2 = (unsigned int *) 0 ;
  int arg3 ;
  void *argp1 = 0 ;
  int res1 = 0 ;
  int val3 ;
  int ecode3 = 0 ;
  LookupResult *result = 0 ;
    
  if(args.Length() != 2) SWIG_exception_fail(SWIG_ERROR, "Illegal number of arguments for _wrap_AudioIndex_lookup.");
    
  res1 = SWIG_ConvertPtr(args.Holder(), &argp1,SWIGTYPE_p_AudioIndex, 0 |  0 );
  if (!SWIG_IsOK(res1)) {
    SWIG_exception_fail(SWIG_ArgError(res1), "in method '" "AudioIndex_lookup" "', argument " "1"" of type '" "AudioIndex *""'");
  }
  arg1 = (AudioIndex *)(argp1);
  {
    if (args[0]->IsArray())
    {
      // Convert into Array
      array = v8::Local<v8::Array>::Cast(args[0]);
  
      length = arg2_dim0;
  
```

So a lot of errors in compilation like

```
asign_wrap_js.cpp: In function 'SwigV8ReturnValue _wrap_AudioIndex_lookup(const SwigV8Arguments&)':
asign_wrap_js.cpp:2350:7: error: 'array' was not declared in this scope
       array = v8::Local<v8::Array>::Cast(args[0]);
       ^
asign_wrap_js.cpp:2352:7: error: 'length' was not declared in this scope
       length = arg2_dim0;
       ^
asign_wrap_js.cpp:2352:16: error: 'arg2_dim0' was not declared in this scope
       length = arg2_dim0;
                ^
asign_wrap_js.cpp:2357:12: error: 'i' was not declared in this scope
       for (i = 0; i < length; i++)
```

I'm sorry, not sure what is the reason the part in parenthesis is ignored.

**Comment by ojwb** [comment]
My guess is the templated types in the list - not helpful that it gets quietly ignored though.

We're lacking a javascript maintainer at the moment, but we're happy to apply clean patches with suitable test cases.

If you're not up to digging in yourself, an input file which causes this and the SWIG command line used to process it would be helpful for anyone wanting to investigate.

**Comment by nshmyrev** [comment]
Thanks, I'll try to look. Example is very simple

```
// Run with swig -javascript -node -DV8_VERSION=0x032872 test.i

%module test
%include <arrays_javascript.i>
%begin %{
    int test(int *a) {
        return 0;
    }
%}
%apply int[] {int *data}
int test(int *data);
```

**Comment by ojwb** [comment]
You don't need/want to use `%begin` there - a top-level `%{` ... `%}` block will just get inserted into the generated wrapper script.

But better to use `%inline` and not have to repeat the code being wrapped:

```
// Run with swig -javascript -node -DV8_VERSION=0x032872 test.i

%module test
%include <arrays_javascript.i>
%apply int[] {int *data}
%inline %{
    int test(int *data) {
        return 0;
    }
%}
```

Anyway, that's not the problem!

I didn't look at the bug with those declarations being ignored, but I had a quick look at the typemaps and they aren't necessary - those variables don't need to live longer than the body of the typemap AFAICS, so we can just declare the variables within the typemap code.

I don't have all the node stuff installed, but I pushed a branch which hopefully works around this bug in this case.  It ought to have a testcase and we ought to actually solve the underlying bug.

**Comment by nshmyrev** [comment]
Thanks, @ojwb , if thats the way to go I'll continue from there and update the patch to make it actually working. I'll also write the test. Your current state is still a problem because _dim0 does not make much sense in node now, it should be array->Length() instead. Leave it, I'll submit a pull request in coming days.

**Comment by ojwb** [comment]
In normal use of this typemap, `_dim0` will be the number of elements in the array - e.g. 10 for something like:

```
int test(int data[10])
```

Your `%apply` doesn't work as the size of the passed C array isn't knowable from just a pointer.  For the input typemaps, `array->Length()` helps you know how much data there is to copy, but you still really need to know how big the passed buffer is or else there's a buffer overflow waiting to happen.

**Comment by nshmyrev** [comment]
Ok, I understood why it does not work, here is the quote from the docs:

Note:

> When declaring a typemap for multiple types, each type must have its own local variable declaration.
> 
> %typemap(in) const std::string *, std::string * (std::string temp) // NO!
> // only std::string * has a local variable
> // const std::string * does not (oops)
> ....
> 
> %typemap(in) const std::string * (std::string temp), std::string * (std::string temp) // Correct
> ....

So it should be 

`%typemap(in, fragment="SWIG_JSCGetIntProperty") int[] (int length = 0, v8::Local<v8::Array> array, v8::Local<v8::Value> jsvalue, int i = 0, int res = 0, $*1_ltype temp), int[ANY]
`

With separate declarations for int[] and int[ANY]. And dim0 works only for int[ANY], not for int[], this case should be handled specially.

**Comment by ojwb** [comment]
Ah yes, that makes sense - so that part isn't a bug at least.

**Comment by wsfulton** [comment]
Fixed in #872

**Comment by nshmyrev** [comment]
Thanks a lot William!
